### ISSUE TITLE
FID does not identify some library functions when "Decompiler Switch Analysis" is checked

### ISSUE THREAD

**Issue by as0ni** [NationalSecurityAgency/ghidra#2877]
For some reason, FID identifies less library functions when "Decompiler Switch Analysis" is checked for Auto Analysis options. 

Here are mem* functions identified when "Decompiler Switch Analysis" is _not_ checked: 
![without_decompiler_switch](https://user-images.githubusercontent.com/8812157/112854413-d262ce80-907b-11eb-8575-4ac3e563637a.png)

Here are mem* functions identified when "Decompiler Switch Analysis"  _is_ checked: 
![with_decompiler_switch](https://user-images.githubusercontent.com/8812157/112854450-dd1d6380-907b-11eb-8193-aa8b4a5fba0c.png)

As you can see, memmove and memset are only identified if "Decompiler Switch Analysis" is not checked. Any idea why? 

This is the sample I analyzed for this test: https://www.virustotal.com/gui/file/595a54f0bbf297041ce259461ae8a12f37fb29e5180705eafb3668b4a491cecc/community

I'm using Version 9.2.2 on Windows.

**Comment by quietust** [comment]
I suspect the specific cases of memset/memmove might be due to them using [Duff's device](https://en.wikipedia.org/wiki/Duff%27s_device) in order to improve performance.

**Comment by as0ni** [comment]
Beginning with 10.1, Ghidra no longer recognizes important library functions like memset, memmove, and fread even if "Decompiler Switch Analysis" is not checked. Can someone look into this? I'm trying to use the latest version for the new SANS FOR710 Advanced RE course, and this is a bit of an issue.

**Comment by emteere** [comment]
Can you please run the script FIDHashCurrentFunction script on the functions that should be memmove, memset, fread and post the output from the script.
Also can you select the bytes from the functions and copy special as a hex string byte string and post them here.

**Comment by emteere** [comment]
The switch analyzer really shouldn't be an issue, unless there is a switch statement in the function that recovers differently than the when the functions were originally analyzed to be added to the FIDB files.  The FID files are dependent on the body of the function being recovered consistently.
It is possibly that the body of the functions have changed recently (better or worse analysis).
There are several other things that could cause the issue, but we'd need to see the functions.

**Comment by as0ni** [comment]
@emteere, see below: 

FIDHashCurrentFunction script output on the functions that should be memmove, memset, fread (in that order):
```
FIDHashCurrentFunction.java> Running...
FIDHashCurrentFunction.java> FID Hash for FUN_140003ec0 at 140003ec0:  FH: 5fe6bd00e261e8ec (177) +121 XH: d49470d197660ecd
FIDHashCurrentFunction.java> Finished!
FIDHashCurrentFunction.java> Running...
FIDHashCurrentFunction.java> FID Hash for FUN_140005880 at 140005880:  FH: 8893ba693ee4add6 (59) +22 XH: 524c7006815c20db
FIDHashCurrentFunction.java> Finished!
FIDHashCurrentFunction.java> Running...
FIDHashCurrentFunction.java> FID Hash for FUN_1400071e4 at 1400071e4:  FH: 928db2cca0136c09 (7) +4 XH: 67d5096e7648ed36
FIDHashCurrentFunction.java> Finished!
```

memmove
`4c 8b d9 4c 8b d2 49 83 f8 10 0f 86 70 00 00 00 49 83 f8 20 76 4a 48 2b d1 73 0f 49 8b c2 49 03 c0 48 3b c8 0f 8c 36 03 00 00 49 81 f8 80 00 00 00 0f 86 69 02 00 00 0f ba 25 19 18 02 00 01 0f 83 ab 01 00 00 49 8b c3 4c 8b df 48 8b f9 49 8b c8 4c 8b c6 49 8b f2 f3 a4 49 8b f0 49 8b fb c3 0f 10 02 41 0f 10 4c 10 f0 0f 11 01 41 0f 11 4c 08 f0 48 8b c1 c3
`
memset
`4c 8b d9 0f b6 d2 49 b9 01 01 01 01 01 01 01 01 4c 0f af ca 49 83 f8 10 0f 86 02 01 00 00 66 49 0f 6e c1 66 0f 60 c0 49 81 f8 80 00 00 00 0f 86 7c 00 00 00 0f ba 25 5c fe 01 00 01 73 22 8b c2 48 8b d7 48 8b f9 49 8b c8 f3 aa 48 8b fa 49 8b c3 c3`

fread
`48 83 ec 38 4c 89 4c 24 20 4d 8b c8 4c 8b c2 48 83 ca ff e8 08 00 00 00 48 83 c4 38 c3
`

**Comment by as0ni** [comment]
@emteere in case it's helpful, here is a side-by-side comparison of memset in 10.0.4 (left) vs. 10.1.1  (right) ("Decompiler Switch Analysis" not checked in both cases). 
![ghira_comparison](https://user-images.githubusercontent.com/8812157/152217992-e4e88b3e-b79e-4650-a2e4-d60f277cd708.png)

**Comment by emteere** [comment]
The bodies of those functions are not the full function bytes.  Just want to make sure the function body is correct.  If you Select->Function do you get more bytes?  There are several conditional jumps to further into the function body.  Can you post all the bytes?

I'd also like to know if the bytes for the function in 10.0.4 differs from 10.1.1?  Can you post both?

And if we aren't getting all the bytes because a switch stmt isn't recovering, can you post all the bytes?
I have them here, I just want to make sure we have the exact same thing, and it is easier than hunting for them :-)

**Comment by as0ni** [comment]
@emteere sorry about that, here you go. 

**memmove 10.0.4:** 
`4c 8b d9 4c 8b d2 49 83 f8 10 0f 86 70 00 00 00 49 83 f8 20 76 4a 48 2b d1 73 0f 49 8b c2 49 03 c0 48 3b c8 0f 8c 36 03 00 00 49 81 f8 80 00 00 00 0f 86 69 02 00 00 0f ba 25 19 18 02 00 01 0f 83 ab 01 00 00 49 8b c3 4c 8b df 48 8b f9 49 8b c8 4c 8b c6 49 8b f2 f3 a4 49 8b f0 49 8b fb c3 0f 10 02 41 0f 10 4c 10 f0 0f 11 01 41 0f 11 4c 08 f0 48 8b c1 c3 48 8b c1 4c 8d 0d b6 c0 ff ff 43 8b 8c 81 57 3f 00 00 49 03 c9 ff e1 0f 10 04 0a 4c 03 c1 48 83 c1 10 41 f6 c3 0f 74 13 0f 28 c8 48 83 e1 f0 0f 10 04 0a 48 83 c1 10 41 0f 11 0b 4c 2b c1 4d 8b c8 49 c1 e9 07 0f 84 88 00 00 00 0f 29 41 f0 4c 3b 0d b1 ff 01 00 76 17 e9 c2 00 00 00 0f 29 41 e0 0f 29 49 f0 0f 10 04 0a 0f 10 4c 0a 10 48 81 c1 80 00 00 00 0f 29 41 80 0f 29 49 90 0f 10 44 0a a0 0f 10 4c 0a b0 49 ff c9 0f 29 41 a0 0f 29 49 b0 0f 10 44 0a c0 0f 10 4c 0a d0 0f 29 41 c0 0f 29 49 d0 0f 10 44 0a e0 0f 10 4c 0a f0 75 ad 0f 29 41 e0 49 83 e0 7f 0f 28 c1 eb 0c 0f 10 04 0a 48 83 c1 10 49 83 e8 10 4d 8b c8 49 c1 e9 04 74 1c 66 66 66 0f 1f 84 00 00 00 00 00 0f 11 41 f0 0f 10 04 0a 48 83 c1 10 49 ff c9 75 ef 49 83 e0 0f 74 0d 49 8d 04 08 0f 10 4c 02 f0 0f 11 48 f0 0f 11 41 f0 49 8b c3 c3 0f 2b 41 e0 0f 2b 49 f0 0f 18 84 0a 00 02 00 00 0f 10 04 0a 0f 10 4c 0a 10 48 81 c1 80 00 00 00 0f 2b 41 80 0f 2b 49 90 0f 10 44 0a a0 0f 10 4c 0a b0 49 ff c9 0f 2b 41 a0 0f 2b 49 b0 0f 10 44 0a c0 0f 10 4c 0a d0 0f 18 84 0a 40 02 00 00 0f 2b 41 c0 0f 2b 49 d0 0f 10 44 0a e0 0f 10 4c 0a f0 75 9d 0f ae f8 e9 38 ff ff ff 49 03 c8 0f 10 44 0a f0 48 83 e9 10 49 83 e8 10 f6 c1 0f 74 17 48 8b c1 48 83 e1 f0 0f 10 c8 0f 10 04 0a 0f 11 08 4c 8b c1 4d 2b c3 4d 8b c8 49 c1 e9 07 74 68 0f 29 01 eb 0d 0f 29 41 10 0f 29 09 0f 10 44 0a f0 0f 10 4c 0a e0 48 81 e9 80 00 00 00 0f 29 41 70 0f 29 49 60 0f 10 44 0a 50 0f 10 4c 0a 40 49 ff c9 0f 29 41 50 0f 29 49 40 0f 10 44 0a 30 0f 10 4c 0a 20 0f 29 41 30 0f 29 49 20 0f 10 44 0a 10 0f 10 0c 0a 75 ae 0f 29 41 10 49 83 e0 7f 0f 28 c1 4d 8b c8 49 c1 e9 04 74 1a 66 66 0f 1f 84 00 00 00 00 00 0f 11 01 48 83 e9 10 0f 10 04 0a 49 ff c9 75 f0 49 83 e0 0f 74 08 41 0f 10 0a 41 0f 11 0b 0f 11 01 49 8b c3 c3`

**memmove 10.1.1:**
`4c 8b d9 4c 8b d2 49 83 f8 10 0f 86 70 00 00 00 49 83 f8 20 76 4a 48 2b d1 73 0f 49 8b c2 49 03 c0 48 3b c8 0f 8c 36 03 00 00 49 81 f8 80 00 00 00 0f 86 69 02 00 00 0f ba 25 19 18 02 00 01 0f 83 ab 01 00 00 49 8b c3 4c 8b df 48 8b f9 49 8b c8 4c 8b c6 49 8b f2 f3 a4 49 8b f0 49 8b fb c3 0f 10 02 41 0f 10 4c 10 f0 0f 11 01 41 0f 11 4c 08 f0 48 8b c1 c3 48 8b c1 4c 8d 0d b6 c0 ff ff 43 8b 8c 81 57 3f 00 00 49 03 c9 ff e1 0f 10 04 0a 4c 03 c1 48 83 c1 10 41 f6 c3 0f 74 13 0f 28 c8 48 83 e1 f0 0f 10 04 0a 48 83 c1 10 41 0f 11 0b 4c 2b c1 4d 8b c8 49 c1 e9 07 0f 84 88 00 00 00 0f 29 41 f0 4c 3b 0d b1 ff 01 00 76 17 e9 c2 00 00 00 0f 29 41 e0 0f 29 49 f0 0f 10 04 0a 0f 10 4c 0a 10 48 81 c1 80 00 00 00 0f 29 41 80 0f 29 49 90 0f 10 44 0a a0 0f 10 4c 0a b0 49 ff c9 0f 29 41 a0 0f 29 49 b0 0f 10 44 0a c0 0f 10 4c 0a d0 0f 29 41 c0 0f 29 49 d0 0f 10 44 0a e0 0f 10 4c 0a f0 75 ad 0f 29 41 e0 49 83 e0 7f 0f 28 c1 eb 0c 0f 10 04 0a 48 83 c1 10 49 83 e8 10 4d 8b c8 49 c1 e9 04 74 1c 66 66 66 0f 1f 84 00 00 00 00 00 0f 11 41 f0 0f 10 04 0a 48 83 c1 10 49 ff c9 75 ef 49 83 e0 0f 74 0d 49 8d 04 08 0f 10 4c 02 f0 0f 11 48 f0 0f 11 41 f0 49 8b c3 c3 0f 2b 41 e0 0f 2b 49 f0 0f 18 84 0a 00 02 00 00 0f 10 04 0a 0f 10 4c 0a 10 48 81 c1 80 00 00 00 0f 2b 41 80 0f 2b 49 90 0f 10 44 0a a0 0f 10 4c 0a b0 49 ff c9 0f 2b 41 a0 0f 2b 49 b0 0f 10 44 0a c0 0f 10 4c 0a d0 0f 18 84 0a 40 02 00 00 0f 2b 41 c0 0f 2b 49 d0 0f 10 44 0a e0 0f 10 4c 0a f0 75 9d 0f ae f8 e9 38 ff ff ff 49 03 c8 0f 10 44 0a f0 48 83 e9 10 49 83 e8 10 f6 c1 0f 74 17 48 8b c1 48 83 e1 f0 0f 10 c8 0f 10 04 0a 0f 11 08 4c 8b c1 4d 2b c3 4d 8b c8 49 c1 e9 07 74 68 0f 29 01 eb 0d 0f 29 41 10 0f 29 09 0f 10 44 0a f0 0f 10 4c 0a e0 48 81 e9 80 00 00 00 0f 29 41 70 0f 29 49 60 0f 10 44 0a 50 0f 10 4c 0a 40 49 ff c9 0f 29 41 50 0f 29 49 40 0f 10 44 0a 30 0f 10 4c 0a 20 0f 29 41 30 0f 29 49 20 0f 10 44 0a 10 0f 10 0c 0a 75 ae 0f 29 41 10 49 83 e0 7f 0f 28 c1 4d 8b c8 49 c1 e9 04 74 1a 66 66 0f 1f 84 00 00 00 00 00 0f 11 01 48 83 e9 10 0f 10 04 0a 49 ff c9 75 f0 49 83 e0 0f 74 08 41 0f 10 0a 41 0f 11 0b 0f 11 01 49 8b c3 c3`

**memset 10.0.4:** 
`4c 8b d9 0f b6 d2 49 b9 01 01 01 01 01 01 01 01 4c 0f af ca 49 83 f8 10 0f 86 02 01 00 00 66 49 0f 6e c1 66 0f 60 c0 49 81 f8 80 00 00 00 0f 86 7c 00 00 00 0f ba 25 5c fe 01 00 01 73 22 8b c2 48 8b d7 48 8b f9 49 8b c8 f3 aa 48 8b fa 49 8b c3 c3 0f 11 01 4c 03 c1 48 83 c1 10 48 83 e1 f0 4c 2b c1 4d 8b c8 49 c1 e9 07 74 36 66 0f 1f 44 00 00 0f 29 01 0f 29 41 10 48 81 c1 80 00 00 00 0f 29 41 a0 0f 29 41 b0 49 ff c9 0f 29 41 c0 0f 29 41 d0 0f 29 41 e0 66 0f 29 41 f0 75 d4 49 83 e0 7f 4d 8b c8 49 c1 e9 04 74 13 0f 1f 80 00 00 00 00 0f 11 01 48 83 c1 10 49 ff c9 75 f4 49 83 e0 0f 74 06 41 0f 11 44 08 f0 49 8b c3 c3 49 8b d1 4c 8d 0d 56 a6 ff ff 43 8b 84 81 5c 59 00 00 4c 03 c8 49 03 c8 49 8b c3 41 ff e1`

**memset 10.1.1:**
`4c 8b d9 0f b6 d2 49 b9 01 01 01 01 01 01 01 01 4c 0f af ca 49 83 f8 10 0f 86 02 01 00 00 66 49 0f 6e c1 66 0f 60 c0 49 81 f8 80 00 00 00 0f 86 7c 00 00 00 0f ba 25 5c fe 01 00 01 73 22 8b c2 48 8b d7 48 8b f9 49 8b c8 f3 aa 48 8b fa 49 8b c3 c3 0f 11 01 4c 03 c1 48 83 c1 10 48 83 e1 f0 4c 2b c1 4d 8b c8 49 c1 e9 07 74 36 66 0f 1f 44 00 00 0f 29 01 0f 29 41 10 48 81 c1 80 00 00 00 0f 29 41 a0 0f 29 41 b0 49 ff c9 0f 29 41 c0 0f 29 41 d0 0f 29 41 e0 66 0f 29 41 f0 75 d4 49 83 e0 7f 4d 8b c8 49 c1 e9 04 74 13 0f 1f 80 00 00 00 00 0f 11 01 48 83 c1 10 49 ff c9 75 f4 49 83 e0 0f 74 06 41 0f 11 44 08 f0 49 8b c3 c3 49 8b d1 4c 8d 0d 56 a6 ff ff 43 8b 84 81 5c 59 00 00 4c 03 c8 49 03 c8 49 8b c3 41 ff e1`

**fread 10.0.4:** 
`48 83 ec 38 4c 89 4c 24 20 4d 8b c8 4c 8b c2 48 83 ca ff e8 08 00 00 00 48 83 c4 38 c3`

**fread 10.1.1:** 
`48 83 ec 38 4c 89 4c 24 20 4d 8b c8 4c 8b c2 48 83 ca ff e8 08 00 00 00 48 83 c4 38 c3`

**Comment by emteere** [comment]
Ok, one more ask, can you please select the bytes of memset from the beginning of memset to the last return instruction, including any intervening instructions that don't disassemble (filler, code not found, any data).

**Comment by as0ni** [comment]
@emteere, here you go in case it's still helpful: 

**memset v10.0.4**
`4c 8b d9 0f b6 d2 49 b9 01 01 01 01 01 01 01 01 4c 0f af ca 49 83 f8 10 0f 86 02 01 00 00 66 49 0f 6e c1 66 0f 60 c0 49 81 f8 80 00 00 00 0f 86 7c 00 00 00 0f ba 25 5c fe 01 00 01 73 22 8b c2 48 8b d7 48 8b f9 49 8b c8 f3 aa 48 8b fa 49 8b c3 c3 66 66 66 66 66 66 0f 1f 84 00 00 00 00 00 0f 11 01 4c 03 c1 48 83 c1 10 48 83 e1 f0 4c 2b c1 4d 8b c8 49 c1 e9 07 74 36 66 0f 1f 44 00 00 0f 29 01 0f 29 41 10 48 81 c1 80 00 00 00 0f 29 41 a0 0f 29 41 b0 49 ff c9 0f 29 41 c0 0f 29 41 d0 0f 29 41 e0 66 0f 29 41 f0 75 d4 49 83 e0 7f 4d 8b c8 49 c1 e9 04 74 13 0f 1f 80 00 00 00 00 0f 11 01 48 83 c1 10 49 ff c9 75 f4 49 83 e0 0f 74 06 41 0f 11 44 08 f0 49 8b c3 c3 ce 59 00 00 cb 59 00 00 f7 59 00 00 c7 59 00 00 d4 59 00 00 e4 59 00 00 f4 59 00 00 c4 59 00 00 fc 59 00 00 d8 59 00 00 10 5a 00 00 00 5a 00 00 d0 59 00 00 e0 59 00 00 f0 59 00 00 c0 59 00 00 18 5a 00 00 49 8b d1 4c 8d 0d 56 a6 ff ff 43 8b 84 81 5c 59 00 00 4c 03 c8 49 03 c8 49 8b c3 41 ff e1`

**memset 10.1.1**
`4c 8b d9 0f b6 d2 49 b9 01 01 01 01 01 01 01 01 4c 0f af ca 49 83 f8 10 0f 86 02 01 00 00 66 49 0f 6e c1 66 0f 60 c0 49 81 f8 80 00 00 00 0f 86 7c 00 00 00 0f ba 25 5c fe 01 00 01 73 22 8b c2 48 8b d7 48 8b f9 49 8b c8 f3 aa 48 8b fa 49 8b c3 c3 66 66 66 66 66 66 0f 1f 84 00 00 00 00 00 0f 11 01 4c 03 c1 48 83 c1 10 48 83 e1 f0 4c 2b c1 4d 8b c8 49 c1 e9 07 74 36 66 0f 1f 44 00 00 0f 29 01 0f 29 41 10 48 81 c1 80 00 00 00 0f 29 41 a0 0f 29 41 b0 49 ff c9 0f 29 41 c0 0f 29 41 d0 0f 29 41 e0 66 0f 29 41 f0 75 d4 49 83 e0 7f 4d 8b c8 49 c1 e9 04 74 13 0f 1f 80 00 00 00 00 0f 11 01 48 83 c1 10 49 ff c9 75 f4 49 83 e0 0f 74 06 41 0f 11 44 08 f0 49 8b c3 c3 ce 59 00 00 cb 59 00 00 f7 59 00 00 c7 59 00 00 d4 59 00 00 e4 59 00 00 f4 59 00 00 c4 59 00 00 fc 59 00 00 d8 59 00 00 10 5a 00 00 00 5a 00 00 d0 59 00 00 e0 59 00 00 f0 59 00 00 c0 59 00 00 18 5a 00 00 49 8b d1 4c 8d 0d 56 a6 ff ff 43 8b 84 81 5c 59 00 00 4c 03 c8 49 03 c8 49 8b c3 41 ff e1`

**Comment by emteere** [comment]
I've managed to duplicate the hash issue and track down all the various pieces.  Thanks for all the info to narrow it down.

The hashes have changed due to subtle changes in the instructions.  We'll have to regenerate the FID files.
There are a few things, like better or worse analysis, instruction changes, or collisions in function names for a dependent function that can cause the FID to not label a function.

**Comment by as0ni** [comment]
Glad to hear the info helped. Thanks for tracking this down so quickly. Makes sense, I guess the FID files need to be generated whenever updates in the analysis result in changes to the instructions.

**Comment by ryanmkurtz** [comment]
This has been fixed for the upcoming 10.1.3 release.  To update your git repo with the latest FID files, do:

```
$ git pull
$ gradle -I gradle/support/fetchDependencies.gradle init
$ gradle prepdev
```

**Comment by as0ni** [comment]
@ryanmkurtz @emteere I finally got around to testing this, and while recent versions of Ghidra can identify functions like memmove, the issue regarding checking "Decompiler Switch Analysis" in the Auto Analysis options remains. That is, if "Decompiler Switch Analysis" is checked (the default) Ghidra will not correctly identify functions like memmove in the sample I referenced (https://www.virustotal.com/gui/file/595a54f0bbf297041ce259461ae8a12f37fb29e5180705eafb3668b4a491cecc/community).
